walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,119 @@
 
 package com.google.sps;
 
-import java.util.Collection;
+import com.google.sps.Event;
+import com.google.sps.Events;
+import com.google.sps.MeetingRequest;
+import com.google.sps.TimeRange;
+import java.util.*;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // If the request meeting is longer than the duration of one day, return an empty list.
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return Collections.emptyList();
+    }
+    // If there is no event on the calendar currently, return whole day as possible time slot.
+    if (events.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    Collection<String> mandatoryAttendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    // If no mandatory attendees or optional attendees, return whole day.
+    if (mandatoryAttendees.isEmpty() && optionalAttendees.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+    // Get unavailable time of mandatory attendees.
+    List<TimeRange> unavailableTimes = getUnavailableTimes(events, mandatoryAttendees);
+    // Merge unavailable intervals with overlaps.
+    List<TimeRange> mergedUnavailableTimes = mergeUnavailableTimes(unavailableTimes);
+    // Get available time slot base on merged unavailable time.
+    List<TimeRange> availableTimes = getAvailableTimes(mergedUnavailableTimes, request.getDuration());
+    // Deal with optional attendees.
+    if (!optionalAttendees.isEmpty()) {
+      List<TimeRange> optionalUnavailableTimes = getUnavailableTimes(events, optionalAttendees);
+      optionalUnavailableTimes.addAll(mergedUnavailableTimes);
+      List<TimeRange> optionalMergedUnavailableTimes = mergeUnavailableTimes(optionalUnavailableTimes);
+      List<TimeRange> combineAvailableTimes = getAvailableTimes(optionalMergedUnavailableTimes, request.getDuration());
+      if (mandatoryAttendees.isEmpty()) {
+        return combineAvailableTimes;
+      } else if (!combineAvailableTimes.isEmpty()) {
+        return combineAvailableTimes;
+      }
+    }
+    return availableTimes;
+  }
+
+  /**
+  * @param  events  a list of existing events.
+  * @param  mandatoryAttendees a list of mandatory attendees for the requested event.
+  * @return a list of maybe overlaping timeRange that at least one mandatory attendee is unavailable
+  */
+  private List<TimeRange> getUnavailableTimes(Collection<Event> events, Collection<String> mandatoryAttendees) {
+    List<TimeRange> unavailableTimes = new ArrayList<>();
+    for (Event e : events) {
+      Set<String> attendees = e.getAttendees();
+      Set<String> intersection = new HashSet<String>(attendees);
+      intersection.retainAll(mandatoryAttendees);
```

Comment:
There's no need to turn these into sets. Collections.disjoint can work on Collections instead of requiring Sets.